src/Artifacts.jl

```diff
@@ -0,0 +1,589 @@
+    
+        if !force && haskey(artifact_dict, name)
+            meta = artifact_dict[name]
+            if !isa(meta, Array)
+                error("Mapping for '$name' within $(artifact_toml) already exists!")
+            elseif any((unpack_platform(x, name, artifact_toml) for x in meta) .== Ref(platform))
+                error("Mapping for '$name'/$(triplet(platform)) within $(artifact_toml) already exists!")
+            end
+        end
+    else
+        artifact_dict = Dict()
+    end
+
+    # Otherwise, the new piece of data we're going to write out is this dict:
+    meta = Dict{String,Any}(
+        "git-tree-sha1" => bytes2hex(hash.bytes),
+    )
+
+    # If we're set to be lazy, then lazy we shall be
+    if lazy
+        meta["lazy"] = true
+    end
+
+    # Integrate download info, if it is given.  We represent the download info as a
+    # vector of dicts, each with its own `url` and `sha256`, since different tarballs can
+    # expand to the same tree hash.
+    if download_info != nothing
+        meta["download"] = [
+            Dict("url" => dl[1],
+                 "sha256" => dl[2],
+            ) for dl in download_info
+        ]
+    end
+
+    if platform == nothing
+        artifact_dict[name] = meta
+    else
+        # Add platform-specific keys to our `meta` dict
+        pack_platform!(meta, platform)
+
+        # Insert this entry into the list of artifacts
+        if !haskey(artifact_dict, name)
+            artifact_dict[name] = [meta]
+        else
+            # Delete any entries that contain identical platforms
+            artifact_dict[name] = filter(
+                x -> unpack_platform(x, name, artifact_toml) != platform,
+                artifact_dict[name]
+            )
+            push!(artifact_dict[name], meta)
+        end
+    end
+
+    # Spit it out onto disk
+    open(artifact_toml, "w") do io
+        TOML.print(io, artifact_dict, sorted=true)
+    end
+    return
+end
+
+
+"""
+    unbind_artifact(name::String, artifact_toml::String; platform = nothing)
+
+Unbind the given `name` from an `Artifact.toml` file.  Silently fails if no such binding
+exists within the file.
+"""
+function unbind_artifact(name::String, artifact_toml::String;
+                         platform::Union{Platform,Nothing} = nothing)
+    artifact_dict = parse_toml(artifact_toml)
+    if !haskey(artifact_dict, name)
+        return
+    end
+
+    if platform == nothing
+        delete!(artifact_dict, name)
+    else
+        artifact_dict[name] = filter(
+            x -> unpack_platform(x, name, artifact_toml) != platform,
+            artifact_dict[name]
+        )
+    end
+
+    open(artifact_toml, "w") do io
+        TOML.print(io, artifact_dict, sorted=true)
+    end
+    return
+end
+
+"""
+    download_artifact(tree_hash::SHA1, tarball_url::String, tarball_hash::String;
+                      verbose::Bool = false)
+
+Download/install an artifact into the artifact store.  Returns `true` on success.
```

Comment:
And on failure?

Reply:
It throws on failure, deep in the `download_verify_unpack()` chain.